### PR TITLE
Re-work the decodeCodec function to return useful information.

### PR DESCRIPTION
**Pull request by dipack95**
We're logging lots of whitespace when encountering codec failures, and these empty trailing spaces which are really annoying to look at.

https://app.datadoghq.com/logs?query=%22Failed%20to%20decode%20codec%22%20&cols=service%2Ccontainer%2C%40location%2C%40req.headers.x-sombra-saas-host%2C%40res.statusCode%2C%40dataSilo.type&event=AgAAAY459maKIXW60gAAAAAAAAAYAAAAAEFZNDU5dEVZQUFESE5wOHR2WmFtX0FBZQAAACQAAAAAMDE4ZTM5ZjgtNmNkNi00Y2E0LWIxMTAtZjc3NWJmMjQ0YzA3&index=%2A&messageDisplay=inline&storage=hot&stream_sort=desc&viz=stream

## Related Issues

- _[none]_

## Security Implications

_[none]_

## System Availability

_[none]_

### CODE REVIEW

**Review by iamtheluckyest — DISMISSED** [review]
I'd advise waiting on more reviews, since I'm no io-ts wizard. But lgtm!

File: src/codecTools/decodeCodec.ts [hunk 1]
```diff
@@ -1,38 +1,46 @@
...
     v,
     fold(
       (errors) =>
         errors.map((error) => {
-          const lastError = error.context[error.context.length - 1];
-          return `${error.context
-            .map((context) => context.key)
-            // Note: lastError.actual is helpful for debugging
-            .join('.')} expected type '${lastError.type.name}'`;
+          const lastCtx = error.context.at(-1);
```
Comment (MadDataScience) [inline src/codecTools/decodeCodec.ts:16]: TIL how to use a negative index in TS!

**Review by giacaglia — DISMISSED** [review]

File: src/codecTools/decodeCodec.ts [hunk 2]
```diff
@@ -48,26 +56,29 @@ export function decodeCodec<TCodec extends t.Any>(
...
+    | undefined = undefined,
 ): t.TypeOf<TCodec> {
   // uncomment the below to view the response pre-decode
   const decoded = codec.decode(
     parse && typeof txt === 'string' ? JSON.parse(txt) : txt,
   );
   // Log errors on failure
-  // eslint-disable-next-line no-underscore-dangle
-  if (decoded._tag === 'Left') {
+  if (isLeft(decoded)) {
```
Comment (MadDataScience) [inline src/codecTools/decodeCodec.ts:68]: What does `isLeft(decoded)` mean?

File: src/codecTools/decodeCodec.ts [hunk 3]
```diff
@@ -1,38 +1,46 @@
...
+ * @param customErrorFromContext - Custom mapper function.
+ * @returns The custom error strings formatted for human consumption.
+ */
+function getCustomErrors<A>(
   v: t.Validation<A>,
   customErrorFromContext: (validationContext: t.Context) => string,
-): string[] =>
-  pipe(
+): string[] {
+  return pipe(
```
Comment (MadDataScience) [inline src/codecTools/decodeCodec.ts:36]: why the change from ` => ` to `{ return ...`? I'm surprised it didn't complain about this.

File: src/codecTools/decodeCodec.ts [hunk 4]
```diff
@@ -1,38 +1,46 @@
...
       (errors) =>
         errors.map((error) => {
-          const lastError = error.context[error.context.length - 1];
-          return `${error.context
-            .map((context) => context.key)
-            // Note: lastError.actual is helpful for debugging
-            .join('.')} expected type '${lastError.type.name}'`;
+          const lastCtx = error.context.at(-1);
+          const fullPath = error.context.map(({ key }) => key).join('.');
+          return `${fullPath} expected type '${lastCtx?.type.name}'`;
```
Comment (MadDataScience) [inline src/codecTools/decodeCodec.ts:18]: Just checking, it looks like nothing is actually changed here apart from just cleaning it up. But this returns the same thing as before, right? Or am I missing something?

**Review by MadDataScience — DISMISSED** [review]
I don't really understand what's changed but I see nothing wrong with it!

**Review by michaelfarrell76 — DISMISSED** [review]
this doesn’t log out the underlying pii by default right?

**Comment by dipack95** [comment]
@michaelfarrell76 Nope!

File: src/codecTools/decodeCodec.ts (hunk 2 above)
Comment (dipack95) [reply src/codecTools/decodeCodec.ts:68]: It's an io-ts thing, where "Left" is used for errors, and "Right" is used for success values. If a response is "Left" it's typically an error, and this helper function provided by the lib just checks to see if the response is an error

File: src/codecTools/decodeCodec.ts (hunk 3 above)
Comment (dipack95) [reply src/codecTools/decodeCodec.ts:36]: So recently I've noticed that if you want docs for a function on hover, then you need them to be a full-fledged function -- considering the style earlier was creating an anonymous function, which ESLint does not enforce docs for, we were missing out on those docs, hence the change

**Review by michaelfarrell76 — APPROVED** [review]

**Review by anotherminh — APPROVED** [review]

File: src/codecTools/decodeCodec.ts (hunk 3 above)
Comment (dipack95) [reply src/codecTools/decodeCodec.ts:36]: > Arrow functions and function functions are actually different in terms of how they handle references to this and arguments.

Yeah we don't typically use `this` references in our code so it was a safe change to make in this instance. Fair feedback though, I misspoke earlier -- I think we should have docs on our functions, and ESLint helps us enforce this if it knows they are functions
